### PR TITLE
[server] properly drain connections on SIGTERM

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

- draing all http and ws endpoints in parallel, not sequential, otherwise first endpoint can block till not drained, preventing draining others
- terminate web socket connections

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 2e16238</samp>

This pull request adds WebSocket support to the server component and improves its cleanup logic. It refactors the `WsExpressHandler` class to use disposables and modifies the `Server` class to handle and dispose of WebSocket connections and wait for all servers and applications to stop gracefully.

</details>

## How to test

- Restart a server while using dashboard. WS should be closed immediately now don't be blocked for 30 secs.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - ak-drain-s6de913315b</li>
	<li><b>🔗 URL</b> - <a href="https://ak-drain-s6de913315b.preview.gitpod-dev.com/workspaces" target="_blank">ak-drain-s6de913315b.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - ak-drain_server_propertly-gha.15630</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/server.ts [hunk 1]
```diff
@@ -344,12 +345,14 @@ export class Server {
...
-        await this.stopServer(this.httpServer);
-        await this.stopServer(this.apiServer);
-        this.disposables.dispose();
+        await Promise.allSettled([
+            this.debugApp.stop(),
+            this.stopServer(this.iamSessionAppServer),
+            this.stopServer(this.monitoringHttpServer),
+            this.stopServer(this.httpServer),
+            this.stopServer(this.apiServer),
+            new Promise(() => this.disposables.dispose()),
```
Comment (geropl) [inline components/server/src/server.ts:354]: Why do we get away here without calling `resolve` explicitly? :thinking: I have the suspicion I will learn something today :laughing:

Comment (akosyakov) [reply components/server/src/server.ts:354]: no, it should have `resolve` 😬 good catch!

File: components/server/src/express/ws-handler.ts [hunk 2]
```diff
@@ -107,4 +114,21 @@ export class WsExpressHandler {
         }
         return pathname === matcher;
     }
+
+    /**
+     * Stop accepting new connections, returns when all connections are closed.
+     */
+    dispose(): Promise<void> {
+        return new Promise((resolve, reject) => {
```
Comment (svenefftinge) [inline components/server/src/express/ws-handler.ts:122]: running this in a promise is missleading as our disposables are not async

**Comment by svenefftinge** [comment]
duplicate of #18548

**Comment by svenefftinge** [comment]
sorry wrong one 😅

**Comment by akosyakov** [comment]
@svenefftinge can we instead take your PR?

**Comment by akosyakov** [comment]
Closing in favor of https://github.com/gitpod-io/gitpod/pull/18550

**Pull request closed without merging**